### PR TITLE
Create zookeeper image with him test.

### PR DESCRIPTION
**Pull request by devcdcc**
Hi, this is my first contribution and i would like to contribute with kafka image for rust, but kafka needs for zookeeper, this is the zookeeper image and tomorrow should i will to publish kafka.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -33,3 +33,5 @@ spectral = "0.6"
 reqwest = { version = "0.10", features = ["blocking"] }
 json = "0.12"
 tokio = { version = "0.2", features = ["macros"] }
+env_logger = "0.7"
+zookeeper = "0.5"
```
Comment (thomaseizinger) [inline Cargo.toml:37]: Missing a newline at the end of the file :)

File: src/images/zookeeper.rs [hunk 2]
```diff
@@ -0,0 +1,88 @@
+use crate::core::Port;
+use crate::{Container, Docker, Image, WaitForMessage};
+use std::collections::HashMap;
+
+const CONTAINER_IDENTIFIER: &str = "bitnami/zookeeper";
```
Comment (thomaseizinger) [inline src/images/zookeeper.rs:5]: Any reason we can't use the official zookeeper image? https://hub.docker.com/_/zookeeper

File: src/images/zookeeper.rs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    arguments: ZookeeperArgs,
+    ports: Option<Vec<Port>>,
+}
+
+impl Default for Zookeeper {
+    fn default() -> Self {
+        Zookeeper {
+            tag: DEFAULT_TAG.to_string(),
+            arguments: ZookeeperArgs {},
+            ports: None,
```
Comment (thomaseizinger) [inline src/images/zookeeper.rs:31]: Does zookeeper have a default port? If yes we can include it here to make usage easier!

File: tests/images.rs [hunk 4]
```diff
@@ -322,3 +329,45 @@ pub fn free_local_port() -> Option<u16> {
...
+        }
+    }
+    let zk_urls = zk_server_urls(host_port);
+    println!("connecting to {}", zk_urls);
+
+    let zk = ZooKeeper::connect(&*zk_urls, Duration::from_secs(15), LoggingWatcher).unwrap();
+    zk.add_listener(|zk_state| println!("New ZkState is {:?}", zk_state));
+
+    let auth = zk.add_auth("digest", vec![1, 2, 3, 4]);
+    println!("authenticated -> {:?}", auth);
```
Comment (thomaseizinger) [inline tests/images.rs:363]: Are these printlns really necessary? I would prefer to keep the tests free of such output!

File: tests/images.rs [hunk 5]
```diff
@@ -322,3 +329,45 @@ pub fn free_local_port() -> Option<u16> {
...
+        match env::var(key) {
+            Ok(val) => val,
+            Err(_) => format!("localhost:{}", port),
+        }
+    }
+    let zk_urls = zk_server_urls(host_port);
+    println!("connecting to {}", zk_urls);
+
+    let zk = ZooKeeper::connect(&*zk_urls, Duration::from_secs(15), LoggingWatcher).unwrap();
+    zk.add_listener(|zk_state| println!("New ZkState is {:?}", zk_state));
```
Comment (thomaseizinger) [inline tests/images.rs:360]: What do we need the listener for in the test?

File: tests/images.rs [hunk 6]
```diff
@@ -322,3 +329,45 @@ pub fn free_local_port() -> Option<u16> {
...
+#[test]
+fn zookeeper_check_connection() {
+    struct LoggingWatcher;
+    impl Watcher for LoggingWatcher {
+        fn handle(&self, e: WatchedEvent) {
+            info!("{:?}", e)
+        }
+    }
+    use std::env;
+    use std::time::Duration;
```
Comment (thomaseizinger) [inline tests/images.rs:342]: These can go at the top of the file.

File: tests/images.rs [hunk 7]
```diff
@@ -322,3 +329,45 @@ pub fn free_local_port() -> Option<u16> {
         .map(|addr| addr.port())
         .ok()
 }
+
+#[test]
+fn zookeeper_check_connection() {
```
Comment (thomaseizinger) [inline tests/images.rs:334]: This test is doing quite a lot. Any chance we can trim does down to the very essentials that are required to assert that things are working?

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
Looks good overall!

I a few comments inline :)

File: src/images/zookeeper.rs (hunk 3 above)
Comment (devcdcc) [reply src/images/zookeeper.rs:31]: Hi, @thomaseizinger  sure. Could i separate zookeeper tests in another file?

File: tests/images.rs [hunk 8]
```diff
@@ -331,43 +331,33 @@ pub fn free_local_port() -> Option<u16> {
...
-    use std::time::Duration;
-    let _ = pretty_env_logger::try_init();
-    let free_local_port = free_local_port().unwrap();
+fn zookeeper_check_existence_directories() {
+    let default_zookeeper_port = 2181;
+    let free_local_port = free_local_port().unwrap_or(default_zookeeper_port);
     let docker = clients::Cli::default();
-    let image = images::zookeeper::Zookeeper::default().with_mapped_port((free_local_port, 2181));
+    let image = images::zookeeper::Zookeeper::default()
+        .with_mapped_port((free_local_port, default_zookeeper_port));
```
Comment (thomaseizinger) [inline tests/images.rs:339]: You don't need to specify this again here, just calling `node.get_host_port` should be sufficient!

File: src/images/zookeeper.rs [hunk 9]
```diff
@@ -0,0 +1,90 @@
+use crate::core::Port;
+use crate::{Container, Docker, Image, WaitForMessage};
+use std::collections::HashMap;
+
+const CONTAINER_IDENTIFIER: &str = "zookeeper";
+const DEFAULT_TAG: &str = "3.5.8";
```
Comment (thomaseizinger) [inline src/images/zookeeper.rs:6]: Why not the latest tag `3.6.2`?

File: src/images/zookeeper.rs [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    arguments: ZookeeperArgs,
+    ports: Option<Vec<Port>>,
+}
+
+impl Default for Zookeeper {
+    fn default() -> Self {
+        Zookeeper {
+            tag: DEFAULT_TAG.to_string(),
+            arguments: ZookeeperArgs {},
+            ports: Some(vec![Port {
```
Comment (thomaseizinger) [inline src/images/zookeeper.rs:32]: So actually, we don't even need this! Sorry to send you on a wild goose chase here :sweat_smile: 

Looking at the source of the docker image (https://hub.docker.com/layers/zookeeper/library/zookeeper/3.6.2/images/sha256-a6ba8cd94070ecd6d30f0c2c0911dc04c7d6c7eef77e35460d8e43cdb7a0bd33?context=explore), it automatically exposes the port `2181`.

That combined with [this](https://github.com/testcontainers/testcontainers-rs/blob/3d80acb27677a5897a5a07a8d5276024a8979d93/src/clients/cli.rs#L101-L109) functionality in testcontainers makes things "just work" if people are using `Container#get_host_port`.

If we specify the port here, we force docker to use 2181 on the host as well which means we can only have one of those images running at the same time.

In conclusion, we should actually not default this here. I guess we still want to keep the `with_mapped_port` functionality but I think that would actually fit better onto the [`RunArgs`](https://github.com/testcontainers/testcontainers-rs/blob/3d80acb27677a5897a5a07a8d5276024a8979d93/src/core/docker.rs#L22) that we introduced recently.

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
Thanks!

Two more things:
1. As commented inline, I think we should actually not set the default port.
2. Please squash all your commits into one after that!

File: src/images/zookeeper.rs (hunk 9 above)
Comment (devcdcc) [reply src/images/zookeeper.rs]: i just thought  in kafka compatibility, i was wrong, we can specify  the version later :)

File: src/images/zookeeper.rs (hunk 10 above)
Comment (devcdcc) [reply src/images/zookeeper.rs]: yeah, I got it, thanks.

**Comment by devcdcc** [comment]
> Thanks!
> 
> Two more things:
> 
> 1. As commented inline, I think we should actually not set the default port.
> 2. Please squash all your commits into one after that!

**Comment by devcdcc** [comment]
@thomaseizinger sorry i did a miss click and close the PR XD
And don't worry about suggestion, this is some of my first rust code.

**Review by thomaseizinger — APPROVED** [review]
Thanks for the contribution @devcdcc !

**Comment by devcdcc** [comment]
Thanks for suggestions @thomaseizinger , i am newbie in rust and i would like to continue contributing to open source projects :D
